src/features/optional-chaining.md

```diff
@@ -0,0 +1,136 @@
+const nameLength = db.user.name.length;
+
+// Less error-prone, but harder to read.
+let nameLength;
+if (db && db.user && db.user.name)
+  nameLength = db.user.name.length;
+```
+
+The above can also be expressed using the ternary operator, which doesn’t exactly help readability:
+
+```js
+const nameLength =
+  (db
+    ? (db.user
+      ? (db.user.name
+        ? db.user.name.length
+        : undefined)
+      : undefined)
+    : undefined);
+```
+
+## Introducing the optional chaining operator { #optional-chaining }
+
+Surely you don’t want to write code like that, so having some alternative is desirable. Some other languages offer an elegant solution to this problem with using a feature called “optional chaining”. According to [a recent spec proposal](https://github.com/tc39/proposal-optional-chaining), “an optional chain is a chain of one or more property accesses and function calls, the first of which begins with the token `?.`.”
+
+Using the new optional chaining operator, we can rewrite the above example as follows:
+
+```js
+// Still checks for errors and is much more readable.
+const nameLength = db?.user?.name?.length;
+```
+
+What happens when `db`, `user`, or `name` is `undefined` or `null`? With the optional chaining operator, JavaScript initializes `nameLength` to `undefined` instead of throwing an error.
+
+Notice that this behavior is also more robust than our check for `if (db && db.user && db.user.name)`. For instance, what if `name` was always guaranteed to be a string? We could change `name?.length` to `name.length`. Then, if `name` were an empty string, we would still get the correct `0` length. That is because the empty string is a falsy value: it behaves like `false` in an `if` clause. The optional chaining operator fixes this common source of bugs.
+
+## Additional syntax forms: calls and dynamic properties
+
+There’s also a version of the operator for calling optional methods:
+
+```js
+// Extends the interface with an optional method, which is present
+// only for admin users.
+const adminOption = db?.user?.validateAdminAndGetPrefs?.().option;
+```
+
+The syntax can feel unexpected, as the `?.()` is the actual operator, which applies to the expression _before_ it.
+
+There’s a third usage of the operator, namely the optional dynamic property access, which is done via `?.[]`. It either returns the value referenced by the argument in the brackets, or `undefined` if there’s no object to get the value from. Here’s a possible use case, following the example from above:
+
+```js
+// Extends the capabilities of the static property access
+// with a dynamically generated property name.
+const optionName = 'optional setting';
+const optionLength = db?.user?.preferences?.[optionName].length;
+```
+
+This last form is also available for optionally indexing arrays, e.g.:
+
+```js
+// If the array is missing the user with the given index,
+// then `userName` gracefully evaluates to `undefined`.
```

Comment:
This is incorrect. We're checking if `usersArray` is nullish, not the element at that index.

Reply:
Thanks, that was an important omission.

Reply:
Also, please check in the 2nd commit, where I added such a section.